src/omp/backend.f90

```diff
@@ -379,18 +412,91 @@ subroutine vecadd_omp(self, a, x, b, y)
     class(field_t), intent(in) :: x
     real(dp), intent(in) :: b
     class(field_t), intent(inout) :: y
+    integer, dimension(3) :: dims
+    integer :: i, j, k, ii
+
+    integer :: nvec, remstart
+
+    if ((x%dir /= y%dir) .or. (x%data_loc /= y%data_loc)) then
+      error stop "Called vector add with incompatible fields"
+    end if
+
+    dims = size(x%data)
+    nvec = dims(1)/SZ
+    remstart = nvec*SZ + 1
 
-    y%data = a*x%data + b*y%data ! fixme
+    !$omp parallel do private(i, ii) collapse(2)
+    do k = 1, dims(3)
+      do j = 1, dims(2)
+        ! Execute inner vectorised loops
+        do ii = 1, nvec
+          !$omp simd
+          do i = 1, SZ
+            y%data(i + (ii - 1)*SZ, j, k) = &
+              a*x%data(i + (ii - 1)*SZ, j, k) + &
+              b*y%data(i + (ii - 1)*SZ, j, k)
+          end do
+          !$omp end simd
+        end do
+
+        ! Remainder loop
+        do i = remstart, dims(1)
+          y%data(i, j, k) = a*x%data(i, j, k) + b*y%data(i, j, k)
+        end do
+      end do
+    end do
+    !$omp end parallel do
 
   end subroutine vecadd_omp
 
   real(dp) function scalar_product_omp(self, x, y) result(s)
+
+    use mpi
+
     implicit none
 
     class(omp_backend_t) :: self
     class(field_t), intent(in) :: x, y
+    integer, dimension(3) :: dims
+    integer :: i, j, k, ii
+    integer :: ierr
+
+    integer :: nvec, remstart
+
+    if ((x%dir /= y%dir) .or. (x%data_loc /= y%data_loc)) then
```

Comment:
`scalar_product` requires the actual dimensions of an array, so making sure that both arrays have a `data_loc /= NONE` is important. I think mesh methods we call here will complain anyways but having a sanity check here can make it easy to debug.

Reply:
Done